### PR TITLE
Add push command

### PR DESCRIPTION
**Pull request by ryancbahan**
### WHY are these changes introduced?

First part of https://github.com/Shopify/internal-cli-foundations/issues/644

### WHAT is this pull request doing?

Adds minimal functionality for `push` command to be hooked up to the `appUpdate` mutation, currently only updating app url. You can test by adding `applicationUrl` to your app toml and modifying the value, then pushing via `config push`.

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/services/app/config/push.ts
Comment (rezaansyed) [inline packages/app/src/cli/services/app/config/push.ts:19]: `const mutation = PushConfig`?

File: packages/app/src/cli/api/graphql/push_config.ts
Comment (ryancbahan) [reply packages/app/src/cli/api/graphql/push_config.ts:17]: sure thing, mind if we handle in a followup since this interfacing is definitely changing?

File: packages/app/src/cli/services/app/config/push.ts
Comment (ryancbahan) [inline packages/app/src/cli/services/app/config/push.ts:18]: this will change as we add `use` and the `--config` flags

**Review by gonzaloriestra — COMMENTED** [review]
Great work! I've added a few comments, but it's working fine.

File: packages/app/src/cli/api/graphql/push_config.ts
Comment (ryancbahan) [reply packages/app/src/cli/api/graphql/push_config.ts:1]: it's only the same temporarily, we'll be adding a bunch more fields soon. we're just waiting on the final config grouping to be aligned upon before adding all of the fields. though if all the new fields end up being nullable, we could reasonably just use one mutation.

File: packages/app/src/cli/models/app/app.ts
Comment (ryancbahan) [reply packages/app/src/cli/models/app/app.ts:12]: we've been waiting on adding grouped tables until the shape is more concretely decided -- similar to why we only added the url field so far. we should be able to update this shortly, though.

Comment (gonzaloriestra) [reply packages/app/src/cli/models/app/app.ts:12]: Ok, I'll sync [my PR for the link command](https://github.com/Shopify/cli/pull/2172) with these values then.

File: packages/app/src/cli/services/app/config/push.ts
Comment (rachel-carvalho) [inline packages/app/src/cli/services/app/config/push.ts:26]: We should also include which field was invalid in the error message. This is what I got with a very long app url:
screenshot with error message without field name

Comment (ryancbahan) [reply packages/app/src/cli/services/app/config/push.ts:33]: is it cool to do this in a followup? i'd like to merge this to get @RyanDJLee unblocked

**Comment by ryancbahan** [comment]
@gonzaloriestra i'm clearing up merge conflicts now -- let me know if this is good to merge so we can unblock build convergence/dependent projects and i can log tickets/fix forward on anything outstanding. if there are any blockers happy to update as well, but hoping to get this merged tomorrow. 

once this has the right approvals/CI/etc, if any of y'all want to merge it before i'm online tomorrow feel free!

**Review by rezaansyed — APPROVED** [review]
LGTM!

**Review by isaacroldan — APPROVED** [review]

File: packages/app/src/cli/models/app/app.ts
Comment (shauns) [inline packages/app/src/cli/models/app/app.ts:12]: I'm not sure on this -- if I load an App current-style then these fields all always be undefined. And if I load new-style then these fields should all always have some value (I think?)

This is tricky as we're dealing with two quite distinct models that underpin such a lot. But if we get the approach wrong we'll feel the debt on this for a very long time. So its a small thing but might have an outsized impact. When we're working new-style for instance, we don't want to drop `if (app.client_id !== undefined)` everywhere when it reality it'd always be set.

I wonder if this might be better done as a [tagged/discriminated union](https://zod.dev/?id=discriminated-unions). If (for instance) we used `name` as the discriminator then:
- if `name` is unset, it can be treated old-style
- if `name` is set, it can be treated new-style, with validation & guarantees on these new fields (and the new sections etc.)

Another alternative could be to have the apploader return something thats an extension of the current output when you're loading new-style.

File: packages/app/src/cli/commands/app/config/push.ts
Comment (shauns) [inline packages/app/src/cli/commands/app/config/push.ts:25]: Maybe: I _think_ this should be `mode: strict`, as we wouldn't want to push a broken app. Report I think tries to fill the blanks if the file is messed up.

File: packages/app/src/cli/services/app/config/push.ts
Comment (RyanDJLee) [reply packages/app/src/cli/services/app/config/push.ts:22]: I think it is a good idea to return a more instructive message, perhaps something like "The repository must be linked to an app to run this command. Refer to [docs]() to learn more." 

However, trying to run `link` for unlinked repositories will break `dev` and `generate` commands support for legacy TOML files / force users to opt-in - although it would not be too strange to only implement this behaviour for `push` since `push` does not make sense out of a linked context.

File: packages/app/src/cli/api/graphql/push_config.ts
Comment (RyanDJLee) [reply packages/app/src/cli/api/graphql/push_config.ts:6]: There was a [thread](https://shopify.slack.com/archives/C04U55Y3ARG/p1686326944774949?thread_ts=1686317504.610199&cid=C04U55Y3ARG) in our project channel regarding conflict resolution although it is out of scope for Summer Editions.

File: packages/app/src/cli/api/graphql/push_config.ts
Comment (RyanDJLee) [reply packages/app/src/cli/api/graphql/push_config.ts:16]: The resolver will return a nice `userErrors`, I just tophatted it.

image

File: packages/app/src/cli/services/app/config/push.ts
Comment (ryancbahan) [reply packages/app/src/cli/services/app/config/push.ts:22]: For sure, we're planning to get placeholder copy in and then jam with Jen on the final copy in a broader take.

**Pull request merged**
